csharp/autobuilder/Semmle.Autobuild/Autobuilder.cs

```diff
@@ -113,46 +119,75 @@ public Autobuilder(IBuildActions actions, AutobuildOptions options)
+            projectsOrSolutionsToBuildLazy = new Lazy<IList<IProjectOrSolution>>(() =>
             {
                 if (options.Solution.Any())
                 {
-                    var ret = new List<ISolution>();
+                    var ret = new List<IProjectOrSolution>();
                     foreach (var solution in options.Solution)
                     {
                         if (actions.FileExists(solution))
                             ret.Add(new Solution(this, solution));
                         else
-                            Log(Severity.Error, "The specified solution file {0} was not found", solution);
+                            Log(Severity.Error, $"The specified solution file {solution} was not found");
                     }
                     return ret;
                 }
 
-                var solutions = GetExtensions(".sln").
-                        Select(s => new Solution(this, s)).
-                        Where(s => s.ProjectCount > 0).
-                        OrderByDescending(s => s.ProjectCount).
-                        ThenBy(s => s.Path.Length).
+                bool FindFiles(string extension, Func<string, ProjectOrSolution> create, out IEnumerable<IProjectOrSolution> files)
+                {
+                    var allFiles = GetExtensions(extension).
+                        Select(p => (ProjectOrSolution: create(p.Item1), DistanceFromRoot: p.Item2)).
+                        Where(p => p.ProjectOrSolution.HasLanguage(this.Options.Language)).
                         ToArray();
 
-                foreach (var sln in solutions)
-                {
-                    Log(Severity.Info, $"Found {sln.Path} with {sln.ProjectCount} {this.Options.Language} projects, version {sln.ToolsVersion}, config {string.Join(" ", sln.Configurations.Select(c => c.FullName))}");
+                    if (allFiles.Length == 0)
+                    {
+                        files = null;
+                        return false;
+                    }
+
+                    if (options.AllSolutions)
+                    {
+                        files = allFiles.Select(p => p.ProjectOrSolution);
+                        return true;
+                    }
+
+                    var firstIsClosest = allFiles.Length > 1 && allFiles[0].DistanceFromRoot < allFiles[1].DistanceFromRoot;
+                    if (allFiles.Length == 1 || firstIsClosest)
+                    {
+                        files = allFiles.Select(p => p.ProjectOrSolution).Take(1);
+                        return true;
+                    }
+
+                    var candidates = allFiles.
+                        Where(f => f.DistanceFromRoot == allFiles[0].DistanceFromRoot).
+                        Select(f => f.ProjectOrSolution);
+                    Log(Severity.Info, $"Found multiple '{extension}' files, giving up: {string.Join(", ", candidates)}.");
+                    files = new IProjectOrSolution[0];
+                    return true;
```

Comment:
This return value is a bit weird - false means carry on and true means stop success or failure.

Reply:
The return value behaves like e.g. the return value of `int.TryParse(string s, out int i)`; the output parameter contains valid data iff the return value is `true`.

Reply:
But actually there is no need to have both the Boolean and the enumerable, so I will rewrite.